File: multiprocessors/examples/05-fork-join/Example8.java
// =================================================================
//
// File: Example8.java
// Author(s):Ramses Aguila Flores A01705291
// David Flores Diaz A01209414
// Speedup =10739 / 5077.7 = 2.1149= 211.49 %
// Description: This file implements the enumeration sort algorithm
// 				using Java's Fork-Join.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================
import java.util.concurrent.RecursiveAction;
import java.util.concurrent.ForkJoinPool;

public class Example8 extends RecursiveAction {
	private static final int SIZE = 100_000;
	private static final int MIN = 1_000;
	private int array[],res[],start,end;

	public Example8 (int array[], int res[], int start, int end){
		this.array = array;
		this.res = res;
		this.start = start;
		this.end = end;	
	}

	public int[] getArray(){
		return res;
	}

	protected void computeDirectly(){
		int k;
                for (int i=start ; i < end ; i++){
                        k = 0;
                        for (int j=0 ; j < SIZE ; j++){
                                if ( array[i] > array[j] ){
                                        k++;
                                }else if( (array[i] == array[j]) && (i < j) ){
                                        k++;
                                }
                        }
                        res[k] = array [i];
                }
	}
	@Override
	protected void compute(){
		if( ( end - start ) <= MIN ){
			computeDirectly();
		}else {
			int mid = (start+end)/2 ;
			invokeAll(new Example8(array,res,start,mid),
				  new Example8(array,res,mid,end));
		}
	}

	public static void main(String args[]) {
		long startTime, stopTime;
		double ms;
		int array[] = new int[SIZE];
		int res[] = new int [SIZE];

		ForkJoinPool pool;
		Utils.randomArray(array);
		Utils.displayArray("before",array);
		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			pool = new ForkJoinPool(Utils.MAXTHREADS);
			pool.invoke(new Example8(array, res, 0, array.length));

			stopTime = System.currentTimeMillis();
			ms += (stopTime - startTime);
		}
		Utils.displayArray("after", res);
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
	}
}

File: multiprocessors/examples/07-tbb/example8.cpp
// =================================================================
//
// File: example8.cpp
// Author:
// A01705291 Ramses Aguila Flores 
// A01209414 David Flores Diaz
//speedup = 381.5452 ms / 91.0088 ms = 419.23 %
// Description: This file contains the code that implements the
//				enumeration sort algorithm using Intel's TBB.
//				To compile: g++ example8.cpp -ltbb
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <cstring>
#include <cmath>
#include <tbb/parallel_for.h>
#include <tbb/blocked_range.h>
#include "utils.h"

const int SIZE = 10000;

using namespace std;
using namespace tbb;

class EnumSort {
private:
	int *array, *res;
public:
	EnumSort(int *a, int *r): array(a) , res(r){}
	EnumSort(EnumSort &x,split): array(x.array) , res(x.res) {}

	void operator() (const blocked_range<int> &r) const{
		int k;
		for (int i= r.begin() ; i != r.end() ; i++){
			k=0;
			for (int j = 0; j < SIZE ; j++){
				if (array[i] > array[j] ) {
					k++;
				}else if(( array[i] == array[j]) && (i < j)){
					k++;
				}
			}
			res[k] = array[i];
		}
	}
};
// place your code here
int main (int argc, char* argv[]){
	int *array, *res;
	double ms;

	array = new int[SIZE];
	random_array(array,SIZE);
	display_array("A",array);

	res = new int[SIZE];

	cout << "Starting..." << endl;
	ms = 0;
	EnumSort obj(array,res);
	for (int i=1 ; i<= N ; i++){
		start_timer();
		parallel_for(blocked_range<int>(0,SIZE),obj);
		ms += stop_timer();
	}
	display_array("Res",res);
	cout << "avg time = " << setprecision(15) << (ms/N) << " ms" << endl;

	delete [] array;
	delete [] res;
	return 0;
}
